File: src/test/java/io/bloco/faker/test_helpers/RegularExpressionMatcher.java
package io.bloco.faker.test_helpers;

import org.hamcrest.Description;
import org.hamcrest.Factory;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeMatcher;

import java.util.regex.Pattern;

public class RegularExpressionMatcher extends TypeSafeMatcher<String> {

    private final Pattern pattern;

    public RegularExpressionMatcher(String pattern) {
        this(Pattern.compile(pattern));
    }
    public RegularExpressionMatcher(Pattern pattern) {
        this.pattern = pattern;
    }

    @Override
    public void describeTo(Description description) {
        description.appendText("matches regular expression ").appendValue(pattern);
    }

    @Override
    public boolean matchesSafely(String item) {
        return pattern.matcher(item).matches();
    }

    @Factory
    public static Matcher matchesPattern(Pattern pattern) {
        return new RegularExpressionMatcher(pattern);
    }
    @Factory
    public static Matcher matchesPattern(String pattern) {
        return new RegularExpressionMatcher(pattern);
    }
}

File: lib/src/test/kotlin/io/bloco/faker/helpers/RegularExpressionMatcher.kt
package io.bloco.faker.helpers

import org.hamcrest.Description
import org.hamcrest.Factory
import org.hamcrest.Matcher
import org.hamcrest.TypeSafeMatcher
import java.util.regex.Pattern

class RegularExpressionMatcher(private val pattern: Pattern) : TypeSafeMatcher<String>() {
    constructor(pattern: String) : this(Pattern.compile(pattern))

    override fun describeTo(description: Description) {
        description.appendText("matches regular expression ").appendValue(pattern)
    }

    public override fun matchesSafely(item: String): Boolean {
        return pattern.matcher(item).matches()
    }

    companion object {
        fun matchesPattern(pattern: Pattern): Matcher<*> {
            return RegularExpressionMatcher(pattern)
        }

        fun matchesPattern(pattern: String): Matcher<String> {
            return RegularExpressionMatcher(pattern)
        }
    }
}